File: Modules/SEDO/EMS/SIE.Web.EMS.SpecialEquipment/RegularInspections/Commands/InputSaveCommand.cs
using SIE.Domain;
using SIE.EMS.SpecialEquipment.RegularInspections;
using SIE.Web.Command;
using System;
using SIE.Domain.Validation;

namespace SIE.Web.EMS.SpecialEquipment.RegularInspections.Commands
{
    /// <summary>
    /// 保存命令
    /// </summary>
    internal class InputSaveCommand : FormSaveCommand
    {
        protected override void DoSave(Entity entity)
        {
            if (entity == null)
            {
                throw new ValidationException("没有数据可以保存。".L10N());
            }
            if (!(entity is RegularInspection))
            {
                throw new ValidationException("该数据不是校验记录数据格式。".L10N());
            }
            var record = entity as RegularInspection;
            RT.Service.Resolve<RegularInspectionController>().SaveRegularInspection(record);
        }
    }
}

File: Modules/SEDO/EMS/SIE.Web.EMS.SpecialEquipment/RegularInspections/Commands/InputSaveCommand.js
SIE.defineCommand('SIE.Web.EMS.SpecialEquipment.RegularInspections.Commands.InputSaveCommand', {
    extend: 'SIE.cmd.FormSave',
    meta: { text: "保存", group: "edit", iconCls: "icon-SaveEntity icon-blue" },

    /**
     * @override 是否可执行
     * @param {} view 
     * @returns {} 
     */
    canExecute: function (view) {
        var current = view.getCurrent();
        if (current) {
            var ApprovalStatu = current.getApprovalStatus();
            var InspectionStatu = current.getInspectionStatus();
            return ApprovalStatu !== SIE.Equipments.Enums.ApprovalStatus.PendingReview.value
                && ApprovalStatu !== SIE.Equipments.Enums.ApprovalStatus.UnderReview.value
                && ApprovalStatu !== SIE.Equipments.Enums.ApprovalStatus.Audited.value
                && InspectionStatu !== SIE.EMS.Enums.InspectionStatus.Calirated.value
                && current.isDirty();
        }
        return this.callParent(arguments);
    },

    canVisible: function (view, source) {
        var current = view.getCurrent();
        if (current) {
            var ApprovalStatu = current.getApprovalStatus();
            var InspectionStatu = current.getInspectionStatus();
            return ApprovalStatu !== SIE.Equipments.Enums.ApprovalStatus.PendingReview.value
                && ApprovalStatu !== SIE.Equipments.Enums.ApprovalStatus.UnderReview.value
                && ApprovalStatu !== SIE.Equipments.Enums.ApprovalStatus.Audited.value
                && InspectionStatu !== SIE.EMS.Enums.InspectionStatus.Calirated.value;
        }
        return this.callParent(arguments);
    },

    /**
     * @override
     * @param {} view 
     * @returns {} 
     */
    onSaved: function (view, res) {
        var me = this;
        var ent = view.getCurrent();

        if (ent.getInspectionStatus() === SIE.EMS.Enums.InspectionStatus.Pending.value) {
            ent.setInspectionStatus(SIE.EMS.Enums.InspectionStatus.Under.value); //改为校验中
        }
        //数据已保存到服务器，修改状态
        ent.markSaved();
        CRT.Event.fire(view.model + "_refresh", ent.data.Id);
        me.onSavedMsg(view, res);
        view.syncCmdState();
    }
});